docs/content/docs/deployment/advanced/failure_enrichers.md

```diff
@@ -0,0 +1,112 @@
+<!--
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+  http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+## Custom failure enrichers
+Flink provides a pluggable interface for users to register their custom logic and enrich failures with extra metadata labels (string kv pairs).
+The goal is to enable developers implement their own failure enrichment plugins to categorize job failures, expose custom metrics, make calls to external notification systems, and more.
+
+FailureEnrichers are triggered every time an exception is reported at runtime by the JobManager.
+Every FailureEnricher may asynchronously return labels associated with the failure that are then exposed via the JobManager's Rest interface (e.g., a 'type:System' label implying the failure is categorized as a system error).
+
+
+### Implement a plugin for your custom enricher
+
+To implement a custom FailureEnricher plugin, you need to:
+
+- Add your own FailureEnricher by implementing the {{< gh_link file="/flink-core/src/main/java/org/apache/flink/core/failure/FailureEnricher.java" name="FailureEnricher" >}} interface.
+
+- Add your own FailureEnricherFactory by implementing the {{< gh_link file="/flink-core/src/main/java/org/apache/flink/core/failure/FailureEnricherFactory.java" name="FailureEnricherFactory" >}} interface.
+
+- Add a service entry. Create a file `META-INF/services/org.apache.flink.core.failure.FailureEnricherFactory` which contains the class name of your failure enricher factory class (see [Java Service Loader](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ServiceLoader.html) docs for more details).
+
+
+Then, create a jar which includes your `FailureEnricher`, `FailureEnricherFactory`, `META-INF/services/` and all the external dependencies.
+Make a directory in `plugins/` of your Flink distribution with an arbitrary name, e.g. "failure-enrichment", and put the jar into this directory.
+See [Flink Plugin]({% link deployment/filesystems/plugins.md %}) for more details.
+
+{{< hint warning >}}
+Note that every FailureEnricher should have defined a set of {{< gh_link file="/flink-core/src/main/java/org/apache/flink/core/failure/FailureEnricher.java" name="output keys" >}} that may be associated with values. This set of keys has to be unique across enrichers otherwise may be ignored.
```

Comment:
> otherwise may be ignored

Otherwise _what_ may be ignored?

Reply:
I don't understand what the implications "ignoring values" is. In practice all that happens if that one enricher would override the label of another one, right? Can we just write that?

Reply:
oh wow all enrichers that clash in any way are just completely dropped if there is a clash. That doesn't seem to match the documented behavior.

Reply:
When enrichers output keys overlap, only the first one is loaded and the rest are dropped with an Error message -- that was the meaning of `This set of keys has to be unique across enrichers otherwise may be ignored.` 
Rephrasing

Reply:
> When enrichers output keys overlap, only the first one is loaded and the rest are dropped with an Error message

That doesn't seem to match `org.apache.flink.runtime.failure.FailureEnricherUtils#filterInvalidEnrichers`. `invalidEnrichers` will contain all enrichers that clashed, and those are all thrown out.

```
        final Set<Class<?>> invalidEnrichers =
                enrichersByKey.entrySet().stream()
                        .filter(entry -> entry.getValue().size() > 1)
                        .flatMap(
                                entry -> {
                                    LOG.warn(
                                            "Following enrichers have have registered duplicate output key [%s] and will be ignored: {}.",
                                            entry.getValue().stream()
                                                    .map(Class::getName)
                                                    .collect(Collectors.joining(", ")));
                                    return entry.getValue().stream();
                                })
                        .collect(Collectors.toSet());
```

`org.apache.flink.runtime.failure.FailureEnricherUtilsTest#testValidatedEnrichersWithInvalidEntries` confirms that, as it uses 3 enrichers, with 2 having clashing keys, and asserting that only 1 is loaded.

Reply:
You are right, that was the initial logic that we then reworked before merging.
Updating README accordingly.